### PR TITLE
do not rename the template in dynamic router

### PR DESCRIPTION
**Pull request by dbu**
$template is in line with the FrameworkBundle TemplateController

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | travis-ci
| Fixed tickets | -
| License       | MIT
| Doc PR        | TODO

together with https://github.com/symfony-cmf/content-bundle/pull/174

### CODE REVIEW

**Comment by ElectricMaxxx** [comment]
Doing so all custom controllers, should call its template parameter `$template`, right? I thought we have some kind of configuration here to be flexible.

**Comment by dbu** [comment]
they have to, if they want parameter automatch. but i think its worth it to avoid the collision with FrameworkBundle. we don't have configuration for that name, only a constant - but a parameter variable name can't use a constant afaik. however, you could write your controller expecting the Request instead, and then get the attribute with the constant...

**Comment by wouterj** [comment]
A couple of remarks here:

 * Why at all do this renaming of `_template` to `contentTemplate`?
 * How can this conflict with the framework's template controller? It'll always require `template`, not `_template`.
 * Why remove `_template` when renaming it to `contentTemplate`? Does it hurt having a duplicated request attribute?
 * If we want to do this change, we should release a new 1.4 version that triggers a deprecation warning when using `contentTemplate`. It should also come with 2 attributes: `template` and `contentTemplate` to allow forward compatibility.

File: UPGRADE-2.0.md [hunk 1]
```diff
@@ -198,3 +198,37 @@
...
+   ```php
+   public function documentAction($contentDocument, $contentTemplate) {
+       ...
+   ```
+
+   **After**
+   ```php
+   public function documentAction($contentDocument, $template) {
+       ...
+   ```
```
Comment (wouterj) [inline UPGRADE-2.0.md]: this file doesn't seem right. The lines 208 - 222 should be removed, right?

**Comment by dbu** [comment]
>   * Why at all do this renaming of |_template| to |contentTemplate|?

having a request attribute called _template triggers an exception by the
framework bundle that _template is reserved for the `@Template` annotation.

>   * How can this conflict with the framework's template controller?
>     It'll always require |template|, not |_template|.

i feel it makes sense to use the same name. its the same concept.

>   * Why remove |_template| when renaming it to |contentTemplate|? Does
>     it hurt having a duplicated request attribute?

yes, having _template triggers an exception.

>   * If we want to do this change, we should release a new 1.4 version
>     that triggers a deprecation warning when using |contentTemplate|. It
>     should also come with 2 attributes: |template| and |contentTemplate|
>     to allow forward compatibility.

putting both into the request attributes seems like a good idea. i don't
know how we could do a deprecation warning however - something in the
core of symfony looks at controller method parameters and tries to match
them with request attribute names.

**Comment by wouterj** [comment]
Let's update this PR to use both `contentTemplate` and `template` in 2.0. We then have more time to see how we can trigger deprecations for the `contentTemplate` (and I believe we should propose to stop throwing an exception when `_template` is present, it doesn't hurt anybody).

**Comment by dbu** [comment]
yeah, i was going to say thats not our fault :-)

File: UPGRADE-2.0.md (hunk 1 above)
Comment (dbu) [inline UPGRADE-2.0.md]: indeed, messed up with copy-pasting it seems.

**Comment by dbu** [comment]
i think this is now ready

File: src/Routing/DynamicRouter.php [hunk 2]
```diff
@@ -41,7 +41,15 @@ class DynamicRouter extends BaseDynamicRouter
...
-    const CONTENT_TEMPLATE = 'contentTemplate';
+    const CONTENT_TEMPLATE = 'template';
+
+    /**
+     * RoutingBundle 1.* put the template into the request attributes as "contentTemplate".
+     *
+     * This name is deprecated in favor of "template" for consistency with
+     * FrameworkBundle and will be removed in RoutingBundle 3.
+     */
+    const BC_CONTENT_TEMPLATE = 'contentTemplate';
```
Comment (wouterj) [inline src/Routing/DynamicRouter.php]: I don't think we have to use a constant for the BC version, it'll only cause more conflicts (people shouldn't start using this constant)

**Comment by wouterj** [comment]
Applied my comment and slightly tweaked the changelog and upgrade guide. Feel free to merge if you agree with my changes.

**Comment by dbu** [comment]
good point, thanks for the cleanup and the merge!

**Pull request merged**